satellite/gracefulexit/endpoint.go

```diff
@@ -0,0 +1,390 @@
+		return Error.Wrap(err)
+	}
+
+	if exitStatus.ExitLoopCompletedAt == nil {
+		err = stream.Send(&pb.SatelliteMessage{Message: &pb.SatelliteMessage_NotReady{NotReady: &pb.NotReady{}}})
+		return Error.Wrap(err)
+	}
+
+	// TODO possibly switch out for custom queue
+	var pending sync.Map
+	pendingLength := func() int {
+		count := 0
+		pending.Range(func(key interface{}, value interface{}) bool {
+			count++
+			return true
+		})
+
+		return count
+	}
+	// this will be 1 until GetIncomplete* methods no longer return values
+	var morePiecesFlag int32 = 1
+
+	var group errgroup.Group
+	group.Go(func() error {
+		ticker := time.NewTicker(endpoint.interval)
+		defer ticker.Stop()
+
+		for range ticker.C {
+			if pendingLength() == 0 {
+				incomplete, err := endpoint.db.GetIncompleteNotFailed(ctx, nodeID, endpoint.config.EndpointBatchSize, 0)
+				if err != nil {
+					return Error.Wrap(err)
+				}
+
+				if len(incomplete) == 0 {
+					incomplete, err = endpoint.db.GetIncompleteFailed(ctx, nodeID, endpoint.config.EndpointMaxFailures+1, endpoint.config.EndpointBatchSize, 0)
+					if err != nil {
+						return Error.Wrap(err)
+					}
+				}
+
+				if len(incomplete) == 0 {
+					endpoint.log.Debug("no more pieces to transfer for node.", zap.String("node ID", nodeID.String()))
+					atomic.StoreInt32(&morePiecesFlag, 0)
+					break
+				}
+
+				for _, inc := range incomplete {
+					pointer, err := endpoint.metainfo.Get(ctx, string(inc.Path))
+					if err != nil {
+						return Error.Wrap(err)
+					}
+					remote := pointer.GetRemote()
+
+					found := false
+					for _, piece := range remote.GetRemotePieces() {
+						if piece.NodeId == nodeID && piece.PieceNum == inc.PieceNum {
+							found = true
+						}
+					}
+					if !found {
+						endpoint.log.Debug("piece no longer held by node.", zap.String("node ID", nodeID.String()), zap.ByteString("path", inc.Path), zap.Int32("piece num", inc.PieceNum))
+						continue
+					}
+
+					redundancy, err := eestream.NewRedundancyStrategyFromProto(pointer.GetRemote().GetRedundancy())
+					if err != nil {
+						return Error.Wrap(err)
+					}
+					pieceSize := eestream.CalcPieceSize(pointer.GetSegmentSize(), redundancy)
+
+					request := overlay.FindStorageNodesRequest{
+						RequestedCount: 1,
+						FreeBandwidth:  pieceSize,
+						FreeDisk:       pieceSize,
+						ExcludedNodes:  []storj.NodeID{nodeID},
+					}
+
+					newNodes, err := endpoint.overlay.FindStorageNodes(ctx, request)
+					if err != nil {
+						return Error.Wrap(err)
+					}
+
+					if len(newNodes) == 0 {
+						return Error.New("could not find a node to transfer this piece to. nodeID %v, path %v, pieceNum %v.", nodeID.String(), zap.ByteString("path", inc.Path), inc.PieceNum)
+					}
+
+					endpoint.log.Debug("found new node for piece transfer.", zap.String("node ID", newNodes[0].Id.String()), zap.ByteString("path", inc.Path), zap.Int32("piece num", inc.PieceNum))
```

Comment:
nit - maybe include the sending node as well as the receiving node in this log message

Reply:
Updated